pkg/dash/dash.go

```diff
@@ -60,21 +62,125 @@ type Options struct {
 }
 
 type Runner struct {
-	dash          *dash
-	pluginManager *plugin.Manager
-	moduleManager *module.Manager
+	dash                   *dash
+	pluginManager          *plugin.Manager
+	moduleManager          *module.Manager
+	actionManager          *action.Manager
+	websocketClientManager *api.WebsocketClientManager
+	loadingComplete        chan bool
 }
 
 func NewRunner(ctx context.Context, logger log.Logger, options Options) (*Runner, error) {
 	ctx = internalLog.WithLoggerContext(ctx, logger)
 
-	r := Runner{}
+	r := Runner{
+		loadingComplete: make(chan bool, 1),
+	}
 
 	if options.Context != "" {
 		logger.With("initial-context", options.Context).Infof("Setting initial context from user flags")
 	}
 
-	logger.Debugf("Loading configuration: %v", options.KubeConfig)
+	actionManger := action.NewManager(logger)
+	r.actionManager = actionManger
+
+	websocketClientManager := api.NewWebsocketClientManager(ctx, r.actionManager)
+	r.websocketClientManager = websocketClientManager
+	go websocketClientManager.Run(ctx)
+
+	listener, err := buildListener()
+	if err != nil {
+		err = fmt.Errorf("failed to create net listener: %w", err)
+		return nil, fmt.Errorf("use OCTANT_LISTENER_ADDR to set host:port: %w", err)
+	}
+
+	// Initialize the API
+	apiService, err := r.initLoadingAPI(ctx, logger, options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to start loading api: %w", err)
+	}
+
+	d, err := newDash(listener, options.Namespace, options.FrontendURL, options.BrowserPath, apiService, logger)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create dash instance: %w", err)
+	}
+
+	if viper.GetBool("disable-open-browser") {
+		d.willOpenBrowser = false
+	}
+
+	d.kubeconfig = apiService.KubeConfig
+	r.dash = d
+
+	return &r, nil
+}
+
+func (r *Runner) Start(ctx context.Context, logger log.Logger, options Options, startupCh, shutdownCh chan bool) error {
+	go func() {
+		if err := r.dash.Run(ctx, startupCh); err != nil {
+			logger.Debugf("running dashboard service: %v", err)
+		}
+		return
+	}()
+
+	go func() {
+		if r.dash != nil {
+			select {
+			case options.KubeConfig = <-findKubeConfig(r.dash.kubeconfig):
+				if options.KubeConfig != "" {
+					logger.Debugf("Loading configuration: %v", options.KubeConfig)
+					r.loadingComplete <- true
+					return
+				}
+			}
+		}
+	}()
+
+	go func() {
+		<-r.loadingComplete
+		apiService, err := r.initAPI(ctx, logger, options)
```

Comment:
Consider merging these into a single go routine. You are blocking on findKubeConfig already. This is the only writer to `r.loadingComplete` which means you could just as well have the contents of the second go routine as a function that gets called.